src/ripple/rpc/ServerHandler.h

```diff
@@ -20,21 +20,200 @@
 #ifndef RIPPLE_RPC_SERVERHANDLER_H_INCLUDED
 #define RIPPLE_RPC_SERVERHANDLER_H_INCLUDED
 
-#include <ripple/basics/BasicConfig.h>
-#include <ripple/beast/utility/Journal.h>
-#include <ripple/core/Config.h>
+#include <ripple/app/main/CollectorManager.h>
 #include <ripple/core/JobQueue.h>
-#include <ripple/resource/ResourceManager.h>
-#include <ripple/rpc/impl/ServerHandlerImp.h>
-#include <ripple/server/Port.h>
-#include <boost/asio/io_service.hpp>
-#include <boost/asio/ip/address.hpp>
-#include <memory>
+#include <ripple/json/Output.h>
+#include <ripple/rpc/RPCHandler.h>
+#include <ripple/rpc/impl/WSInfoSub.h>
+#include <ripple/server/Server.h>
+#include <ripple/server/Session.h>
+#include <ripple/server/WSSession.h>
+#include <boost/beast/core/tcp_stream.hpp>
+#include <boost/beast/ssl/ssl_stream.hpp>
+#include <boost/utility/string_view.hpp>
+#include <condition_variable>
+#include <map>
+#include <mutex>
 #include <vector>
 
 namespace ripple {
 
-using ServerHandler = ServerHandlerImp;
+inline bool
+operator<(Port const& lhs, Port const& rhs)
+{
+    return lhs.name < rhs.name;
+}
+
+class ServerHandler
+{
+public:
+    struct Setup
+    {
+        explicit Setup() = default;
+
+        std::vector<Port> ports;
+
+        // Memberspace
+        struct client_t
+        {
+            explicit client_t() = default;
+
+            bool secure = false;
+            std::string ip;
+            std::uint16_t port = 0;
+            std::string user;
+            std::string password;
+            std::string admin_user;
+            std::string admin_password;
+        };
+
+        // Configuration when acting in client role
+        client_t client;
+
+        // Configuration for the Overlay
+        struct overlay_t
+        {
+            explicit overlay_t() = default;
+
+            boost::asio::ip::address ip;
+            std::uint16_t port = 0;
+        };
+
+        overlay_t overlay;
+
+        void
+        makeContexts();
+    };
+
+private:
+    using socket_type = boost::beast::tcp_stream;
+    using stream_type = boost::beast::ssl_stream<socket_type>;
+
+    Application& app_;
+    Resource::Manager& m_resourceManager;
+    beast::Journal m_journal;
+    NetworkOPs& m_networkOPs;
+    std::unique_ptr<Server> m_server;
+    Setup setup_;
+    JobQueue& m_jobQueue;
+    beast::insight::Counter rpc_requests_;
+    beast::insight::Event rpc_size_;
+    beast::insight::Event rpc_time_;
+    std::mutex mutex_;
+    std::condition_variable condition_;
+    bool stopped_{false};
+    std::map<std::reference_wrapper<Port const>, int> count_;
+
+    // A private type used to restrict access to the ServerHandler constructor.
+    struct ServerHandlerCreator
+    {
+        explicit ServerHandlerCreator() = default;
+    };
+
+    // Friend declaration that allows make_ServerHandler to access the
+    // private type that restricts access to the ServerHandler ctor.
+    friend std::unique_ptr<ServerHandler>
+    make_ServerHandler(
```

Comment:
Can you please point me to resources about this `make_XYX` design pattern? I've read about the factory design pattern and I'm trying to understand it better.

1. Why not make this function a member of the class, instead of declaring it as a `friend`?
2. Why do we need this function? It appears to be calling a constructor of the class (without any additional preprocessing). Can't we make the said constructor public, thereby reducing the number of indirections.

I'm new to this design pattern, I'd like to learn more about it.

Reply:
> 1. Why not make this function a member of the class, instead of declaring it as a `friend`?

It could be a `static` member of the class.  I'm not sure why a free function was preferred over a `static` member in this case.  It's _probably_ for uniformity.  There are quite a number of other `make_` factory functions around the code base.  They are easier for developers to use and think about if they have similar interfaces and behavior.

> 2. Why do we need this function?

Factory functions can have a number of different purposes.  In this case it's because we want to prevent anyone from constructing a `ServerHandler` that is not held by a `std::unique_ptr<ServerHandler>`.  If we made the `ServerHandler` constructor `public`, then anyone could create one.  So, in this case, we give the factory the permission to allow the construction and trust that the factory will do the right thing with the `unique_ptr`.  And we don't have to trust very hard, since the implementation for `make_ServerHandler()` is in the implementation file for `ServerHandler`.

Factory functions can be used for other purposes as well.  For example, the Meyer's Singleton is (in effect) a factory function: https://en.wikipedia.org/wiki/Singleton_pattern

The (ancient) Gang of Four Design Patterns book has a chapter on Factory Method.

Factory functions really just solve the problem that sometimes a simple constructor can't do quite enough.  The factory function can do the extra work that is required.

Reply:
hmm okay. I'll check out these resources 👍